### PR TITLE
SCC-175: Add authentication and authorization

### PR DESCRIPTION
**Pull request by thisisstephenbetts**
Use isso.nypl.org to provide OAuth for the admin app. Internally the app refers to a whitelist of users for authorization.

The log lines should be cleaned up, but that will happen across the app when we use a proper logging module.

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -30,8 +37,99 @@ app.set('views', viewsPath);
 app.set('port', process.env.PORT || appConfig.port);
 // Set the CookieParser middleware
 app.use(cookieParser());
+// Set Global publicKey
+app.set('nyplPublicKey', appConfig.publicKey);
+app.use(session({ secret: '2aad5e41c30cab49515b6dbb903c01b1e0bf590667b25b60f534043c5b554e37' }));
```
Comment (ktp242) [inline server.js]: Should we have this secret in the repo? What is it for?

**Review by ktp242 — APPROVED** [review]

**Comment by rhernand3z** [comment]
This would address #6

File: server.js [hunk 2]
```diff
@@ -30,8 +37,100 @@ app.set('views', viewsPath);
...
+app.use(session({ secret: process.env.SESSION_SECRET || uuidv4() }));
 // Sets the server path to /dist
 app.use(express.static(distPath));
+// Use passport middleware for authentication
+app.use(passport.initialize());
+// use passport sessions
+app.use(passport.session());
+
+// Set up list of authorized users
+let authorized_users = undefined;
```
Comment (rhernand3z) [inline server.js]: As per our code review session, this should be a middleware function added to server/auth/

Using `req.authorizedUsers` to assign a global object to the express app instance.

File: server.js [hunk 3]
```diff
@@ -98,3 +197,9 @@ if (!isProduction) {
 		);
   });
 }
+
+function emailAuthorized(email) {
```
Comment (rhernand3z) [inline server.js]: As per code review session, this should be moved as an exported function in /server/auth

One thing I forgot to mention, it should return a boolean and not throw so when called:

```
if (!isEmailAuthorized(user.email) {
  res.redirect(401, '/error');
}
next();
```

I believe in Node, when you throw, the server cannot handle passing status codes and rendering view layer templates.

File: server.js (hunk 2 above)
Comment (thisisstephenbetts) [reply server.js]: Thanks for this. I'm now making `user` a variable on the req object. As `authorizedUsers` is only seen in the module we shouldn't need to expose it, so having moved it as you suggested, I think we can love it as a module variable. You'll see in the updated PR.

File: server.js (hunk 3 above)
Comment (thisisstephenbetts) [reply server.js]: This is great — incorporating that now. The one place where I do still do a throw is in the auth module, if the `authorizedUsers` list is not a real array, which means that we couldn't download the file or it's malformed. In that case express returns a fairly ugly 500, but as it's a proper error, which should happen exceedingly rarely (and we want app to stop working under those circumstances) that approach seems okay.

File: config/appConfig.js [hunk 4]
```diff
@@ -14,13 +14,15 @@ export default {
     api: process.env.AWS_SQS_API_URL || 'https://sqs.us-east-1.amazonaws.com/224280085904/scsb-item-update-app'
   },
   publicKey:
-		'-----BEGIN PUBLIC KEY-----\n' +
-		'MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAtmIMVmUnbwr/65MYmLGJ\n' +
```
Comment (thisisstephenbetts) [inline config/appConfig.js:18]: This is a different key from that given in https://isso.nypl.org/oauth/key so I updated it

File: package.json [hunk 5]
```diff
@@ -43,22 +43,28 @@
...
+    "connect-ensure-login": "^0.1.1",
     "cookie-parser": "1.4.3",
     "css-loader": "0.28.7",
     "ejs": "2.5.5",
     "express": "4.13.1",
+    "express-session": "^1.15.6",
     "extract-text-webpack-plugin": "2.1.2",
     "immutable": "3.8.1",
     "jsonwebtoken": "7.1.9",
+    "jwt-decode": "^2.2.0",
```
Comment (rhernand3z) [inline package.json]: Can we remove jwt-decode? I think you are using jsonwebtoken?

**Review by rhernand3z — APPROVED** [review]
I think it's a good first iteration, we can handle the React Auth layer next.

File: package.json (hunk 5 above)
Comment (thisisstephenbetts) [reply package.json]: D'oh. Yeah, of course. My bad. Good catch!

**Comment by thisisstephenbetts** [comment]
Thanks Rafael. Removed that line. Will merge.

**Pull request merged**